process/tls_parser.cpp

```diff
@@ -54,291 +50,481 @@ uint64_t quic_get_variable_length(uint8_t *start, uint64_t &offset)
+bool TLSParser::is_grease_value(uint16_t val)
 {
-   if (val != 0 && !(val & ~(0xFAFA)) && ((0x00FF & val) == (val >> 8))) {
-      return true;
-   }
-   return false;
+    return val != 0 && !(val & ~(0xFAFA)) && ((0x00FF & val) == (val >> 8));
 }
 
-void TLSParser::tls_get_quic_user_agent(TLSData &data, char *buffer, size_t buffer_size)
+bool TLSParser::parse_tls(const uint8_t* packet, uint32_t length)
 {
-   // compute end of quic_transport_parameters
-   const uint16_t quic_transport_params_len = ntohs(*(uint16_t *) data.start);
-   const uint8_t *quic_transport_params_end = data.start + quic_transport_params_len
-     + sizeof(quic_transport_params_len);
+    return parse(packet, length, false);
+}
 
-   if (quic_transport_params_end > data.end) {
-      return;
-   }
+bool TLSParser::parse_quic_tls(const uint8_t* packet, uint32_t length)
+{
+    return parse(packet, length, true);
+}
 
-   uint64_t offset = 0;
-   uint64_t param  = 0;
-   uint64_t length = 0;
-
-   while (data.start + offset < quic_transport_params_end) {
-      param  = quic_get_variable_length((uint8_t *) data.start, offset);
-      length = quic_get_variable_length((uint8_t *) data.start, offset);
-      if (param == TLS_EXT_GOOGLE_USER_AGENT) {
-         if (length + (size_t) 1 > buffer_size) {
-            length = buffer_size - 1;
-         }
-         memcpy(buffer, data.start + offset, length);
-         buffer[length] = 0;
-         data.obejcts_parsed++;
-      }
-      offset += length;
-   }
-   return;
+bool TLSParser::parse(const uint8_t* packet, uint32_t length, bool is_quic)
+{
+    m_packet_data = packet;
+    m_packet_length = length;
+    clear_parsed_data();
+
+    if (!parse_tls_header(is_quic)){
+        return false;
+    }
+    if (!parse_tls_handshake()){
+        return false;
+    }
+
+    if (m_handshake->type != TLS_HANDSHAKE_CLIENT_HELLO
+        && m_handshake->type != TLS_HANDSHAKE_SERVER_HELLO) {
+        return false;
+    }
+
+    if (!parse_session_id()) {
+        return false;
+    }
+    if (!parse_cipher_suites()){
+        return false;
+    }
+    if (!parse_compression_methods()){
+        return false;
+    }
+    return true;
 }
 
-void TLSParser::tls_get_server_name(TLSData &data, char *buffer, size_t buffer_size)
+bool TLSParser::parse_tls_header(bool is_quic) noexcept
 {
-   uint16_t list_len       = ntohs(*(uint16_t *) data.start);
-   uint16_t offset         = sizeof(list_len);
-   const uint8_t *list_end = data.start + list_len + offset;
-   size_t buff_offset      = 0;
+    if (is_quic){
+        m_header_section_size = 0;
+        return true;
+    }
+    const auto* tls_header = reinterpret_cast<const TLSHeader*>(m_packet_data);
+
+    if (sizeof(TLSHeader) > m_packet_length) {
+        return false;
+    }
+    if (tls_header == nullptr){
+        return false;
+    }
+    if (tls_header->type != TLS_HANDSHAKE){
+        return false;
+    }
+    if (tls_header->version.major != 3 || tls_header->version.minor > 3){
```

Comment:
Sice se to formatovani zde automaticky nekontroluje, ale i presto zde plati [coding style](https://github.com/CESNET/Nemea-Framework/blob/master/libtrap/CodingStyle.md). 
Jde hlavne o mezery za podminkami V celem tomto MR je to hodne random.